File: src/main/java/de/whitefrog/neobase/model/rest/FieldList.java
package de.whitefrog.neobase.model.rest;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import de.whitefrog.neobase.model.Base;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;

/**
* Created by jochen on 12/01/16.
*/
@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY,
  getterVisibility = JsonAutoDetect.Visibility.NONE,
  setterVisibility = JsonAutoDetect.Visibility.NONE)
public class FieldList extends HashSet<QueryField> {
  public FieldList() {
    super();
  }

  public static FieldList create(QueryField... fields) {
    FieldList list = new FieldList();
    list.addAll(Arrays.asList(fields));
    return list;
  }
  public static FieldList parseFields(String... fields) {
    return parseFields(Arrays.asList(fields), false);
  }
  public static FieldList parseFields(List<String> fields) {
    return parseFields(fields, false);
  }
  public static FieldList parseFields(List<String> fields, boolean addAll) {
    FieldList fieldList = new FieldList();
    
    for(String field: fields) {
      if(field.contains(".")) {
        String fieldName = field.substring(0, field.indexOf("."));
        if(fieldList.containsField(fieldName)) {
          fieldList.get(fieldName).subFields(new QueryField(field.substring(field.indexOf(".") + 1), addAll));
          continue;
        }
      } else if(field.startsWith("[")) {
        // assuming sth like user.[name;login]
        return parseFields(field.substring(1, field.length() - 1).split(";"));
      }
      QueryField queryField = new QueryField(field, addAll);
      if(addAll) queryField.subFields(new QueryField(Base.AllFields));
      fieldList.add(queryField);
    }

    if(addAll) fieldList.add(new QueryField(Base.AllFields));
    return fieldList;
  }

  public boolean containsField(String name) {
    for(QueryField descriptor: this) {
      if(descriptor.field().equals(name)) return true;
    }
    return false;
  }

  public QueryField get(String name) {
    for(QueryField descriptor: this) {
      if(descriptor.field().equals(name)) return descriptor;
    }
    return null;
  }

}

File: base/src/main/kotlin/de/whitefrog/frogr/model/rest/FieldList.kt
package de.whitefrog.frogr.model.rest

import com.fasterxml.jackson.annotation.JsonAutoDetect
import de.whitefrog.frogr.model.Entity

import java.util.Arrays
import java.util.HashSet

/**
 * Field list used in rest queries. Contains multiple QueryField instances.
 */
@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY, getterVisibility = JsonAutoDetect.Visibility.NONE, setterVisibility = JsonAutoDetect.Visibility.NONE)
class FieldList : HashSet<QueryField>() {

  fun containsField(name: String): Boolean {
    return this.any { it.field() == name }
  }

  operator fun get(name: String): QueryField? {
    return this.firstOrNull { it.field() == name }
  }

  companion object {
    @JvmField
    var All = FieldList.parseFields(Entity.AllFields)

    @JvmStatic
    fun create(vararg fields: QueryField): FieldList {
      val list = FieldList()
      list.addAll(Arrays.asList(*fields))
      return list
    }

    @JvmStatic
    fun parseFields(vararg fields: String): FieldList {
      return parseFields(Arrays.asList(*fields), false)
    }

    @JvmStatic @JvmOverloads 
    fun parseFields(fields: List<String>, addAll: Boolean = false): FieldList {
      val fieldList = FieldList()

      for (field in fields) {
        if (field.contains(".")) {
          val fieldName = field.substring(0, field.indexOf("."))
          if (fieldList.containsField(fieldName)) {
            fieldList[fieldName]!!.subFields(QueryField(field.substring(field.indexOf(".") + 1), addAll))
            continue
          }
        } else if (field.startsWith("[")) {
          // assuming sth like user.[name;login]
          return parseFields(*field.substring(1, field.length - 1)
            .split(";".toRegex()).dropLastWhile(String::isEmpty).toTypedArray())
        }
        val queryField = QueryField(field, addAll)
        if (addAll) queryField.subFields(QueryField(Entity.AllFields))
        fieldList.add(queryField)
      }

      if (addAll) fieldList.add(QueryField(Entity.AllFields))
      return fieldList
    }
  }

}
